flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TwoInputStreamOperatorFactory.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.operators;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * A factory to create {@link TwoInputStreamOperator}.
+ *
+ * @param <IN1> The input type of the operator
+ * @param <IN2> The input type of the operator
+ * @param <OUT> The output type of the operator
+ */
+@PublicEvolving
```

Comment:
It should not be possible to extend an `@Experimental` component with a higher stability guarantee. Currently, the project misses a test to cover that behavior.
Unfortunately, this means we need to make the `StreamOperatorFactory` `@PublicEvolving`.

@pnowojski wdyt?

Reply:
+1 from my side for this. It's been `@Experimental` for long enough.

Note that just as for `StreamOperator`, for `StreamOperatorFactory` we will never intend to make it fully `@Public`.

Reply:
@fapaul Thanks for your review. As for this pr, shall we just remove `@PublicEvolving` from `TwoInputStreamOperatorFactory` and `ConnectedStreams`#`transform`(xx, xx, `TwoInputStreamOperatorFactory`)?

After checking [code](https://github.com/apache/flink/blob/d7cf2c10f8d4fba81173854cbd8be27c657c7c7f/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java#L1180), I think  it has the same stability issue too. 

> Currently, the project misses a test to cover that behavior.

As you mentioned above.  Maybe we can create a new jira ticket to address this separately??

Reply:
@shouweikun I'd propose marking `StreamOperatorFactory` as `@PublicEvolving` but please do it as a separate commit. All your other changes are good as they are.